lib/galaxy/jobs/runners/arc.py

```diff
@@ -0,0 +1,593 @@
+import os, sys
+import json
+import logging
+import time
+from datetime import datetime
+import shutil
+
+import requests
+from galaxy.authnz.util import provider_name_to_backend
+
+from galaxy import model
+from galaxy.jobs.runners import (
+    AsynchronousJobRunner,
+    AsynchronousJobState
+)
+from galaxy.util import unicodify
+
+from xml.etree.ElementTree import Element, SubElement, tostring, fromstring
+
+from pyarcrest.common import getNullLogger
+from pyarcrest.arc import ARCJob, ARCRest, ARCRest_1_1
+from pyarcrest.errors import (ARCError, ARCHTTPError)
+
+log = logging.getLogger(__name__)
+
+__all__ = ('ArcRESTJobRunner', )
+
+class Arc:
+    """
+    API parameters
+    """
+
+    def __init__(self):
+
+        self.cluster = ""
+        self.job_mapping = {}
+
+        self.ARC_STATE_MAPPING = {
+            "ACCEPTING": "Accepted",
+            "Accepted": "Accepted",
+            "ACCEPTED": "Accepted",
+            "PREPARING": "Preparing",
+            "PREPARED": "Preparing",
+            "SUBMITTING": "Submitting",
+            "QUEUING": "Queuing",
+            "RUNNING": "Running",
+            "HELD": "Hold",
+            "EXITINGLRMS": "Running",
+            "OTHER": "Other",
+            "EXECUTED": "Running",
+            "FINISHING": "Finishing",
+            "FINISHED": "Finished",
+            "FAILED": "Failed",
+            "KILLING": "Killing",
+            "KILLED": "Killed",
+            "WIPED": "Deleted",
+            "None": "Failed",
+            "Job not found": "Failed"
+        }
+
+        
+    def set_job_cluster(self,cluster):
+        self.cluster = cluster
+
+
+class ArcRESTJobRunner(AsynchronousJobRunner):
+    """
+    Job runner backed by a finite pool of worker threads. FIFO scheduling
+    """
+    runner_name = "ArcRESTJobRunner"
+
+    def __init__(self, app, nworkers, **kwargs):
+        """ 1: Get runner_param_specs from job_conf.xml
+            2: Initialise job runner parent object
+            3: Start the worker and monitor threads
+        """
+
+        # Start the job runner parent object
+        super(ArcRESTJobRunner,self).__init__(app, nworkers, **kwargs)
+
+        self.arc = Arc()
+        self.arcjob = ARCJob()
+        self.provider_backend = provider_name_to_backend("wlcg")
+        #self.arc_url = None
+                
+        
+        """ Following methods starts threads.
+        These methods invoke threading.Thread(name,target)
+        which in turn invokes methods monitor() and run_next().
+        """
+        self._init_monitor_thread()
+        self._init_worker_threads()
```

Comment:
These methods are called automatically on galaxy startup; the logic goes like this: app.py > JobManager().start() > JobHandler().start() > DefaultJobDisplatcher().start() > for each job runner runner.start() where both methods are called [here](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/jobs/runners/__init__.py#L112). Unless this is a special case (which I missed)?

Reply:
I did not review/check if this is necessary, I just followed https://docs.galaxyproject.org/en/master/dev/build_a_job_runner.html 
Maybe it is outdated?

Reply:
Yeah, that part of the documentation can be interpreted both ways (i.e., this is what the runner does vs. this is the code that should be added). I'd leave it out - it should work: all runners start those threads, which is why it's handled in the base runner.